[Bookings] Update type filter logic

irfano opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
Closes WOOMOB-1462

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR implements the logic in `BookingFilterListViewModel` to display the selected type and includes some refactoring.

### Test Steps
<!-- Describe how to test your changes. Include only what’s needed for the reviewer to validate the behavior.
For new features, outline the main user flow or goal rather than every tap or click — the reviewer should be able to complete the flow naturally.
If applicable, mention key devices, scenarios, or edge cases to verify. -->
1. Log in with a CIAB store.
2. Go to Bookings.
3. Go to the All tab.
4. Tap the Filters button.
5. Select Booking type.
6. Select a type.
7. Navigate back.
8. Select Booking type.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
|Before|After|
|-|-|
|<img width="360" src="https://github.com/user-attachments/assets/22730559-ddc8-4817-9107-5a08574bf463" />|<img width="360" src="https://github.com/user-attachments/assets/efe03abb-5faa-4420-8eaf-61e9ce9d64b1" />|

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

irfano committed:
Convert BookingType to a data class

Converts the `BookingType` sealed interface into a data class containing a `Type` enum.

irfano committed:
Move extension property out of companion object

Moved the `BookingType.titleRes` extension property out of the `BookingTypeFilterUiState` companion object to the top level of the file.

irfano committed:
Use subtitle object for booking filter items

Replaced the `value` string in `BookingFilterListItem` with a `BookingFilterListItemSubtitle` data class. This change allows the subtitle to be represented by either a direct string or a string resource ID, providing more flexibility in the UI.

irfano committed:
Display selected booking type filter

Adds a subtitle to the booking filter list to show the currently selected booking type.

irfano committed:
Refactor bookings filter selection logic

irfano committed:
Rename `filterValue` to `remoteValue` in `BookingType`

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/filter/BookingFilterListItem.kt:
```diff
@@ -7,6 +7,9 @@ import androidx.annotation.StringRes
 data class BookingFilterListItem(
     @StringRes val title: Int,
-    val value: String? = null,
+    val subtitle: BookingFilterListItemSubtitle? = null,
```
I renamed `value` to `subtitle` since `BookingFilterListItem` is a UI model, and `subtitle` better reflects its purpose.

AdamGrzybkowski replied:
np: but I'm not sure subtitle is a better name. A subtitle would usually be a text below the title, but this model is used  for the Root booking filter page, where we have the label and the value (currently selected filter). 
<img width="311" height="65" alt="Screenshot 2025-11-05 at 10 01 49" src="https://github.com/user-attachments/assets/0f9656ca-8d20-4c0d-8243-d8238eb4f1e2" />

This would make more sense in a UI like this:
<img width="447" height="78" alt="Screenshot 2025-11-05 at 10 02 22" src="https://github.com/user-attachments/assets/71c41da0-085b-4bb3-acaa-b4a641ee7995" />

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/filter/BookingFilterListItem.kt:
```diff
@@ -7,6 +7,9 @@ import androidx.annotation.StringRes
     val onClick: () -> Unit = {}
-)
+) {
+    data class BookingFilterListItemSubtitle(val valueString: String? = null, @StringRes val valueRes: Int? = null)
```
The `subtitle` sometimes comes from a remote source, such as the customer name, and other times from a string resource, such as “Service.” That’s why I created the `BookingFilterListItemSubtitle` model.

AdamGrzybkowski replied:
💡 Would it make sense to either:
- Make this a sealed class with one non-nullable value, or...
- Keep it like that, but use `UiString` instead. 

The above allows us to call `BookingFilterListItemSubtitle()`. That is not a correct "state". Using the above options would prevent that.

irfano commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/bookings/BookingsFilterOption.kt:
I applied @AdamGrzybkowski's [suggestion](https://github.com/woocommerce/woocommerce-android/pull/14858#discussion_r2480665772) to `BookingType`.

irfano committed:
Merge branch 'trunk' into issue/WOOMOB-1465-update-bookings-type-filter-logic

irfano committed:
Merge branch 'trunk' into issue/WOOMOB-1465-update-bookings-type-filter-logic

AdamGrzybkowski commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/filter/type/BookingTypeFilterUiState.kt:
```diff
@@ -24,7 +24,7 @@ data class BookingTypeFilterUiState(
 }
 
-val BookingType.filterValue: String?
+val BookingType.remoteValue: String?
```
If this is strictly a value we would use to send via REST API, I would create this somewhere else, not in the `...UiState`.

BTW. Is this property used?

irfano replied:
Actually, type filter hasn't been implemented on the backend yet. Besides, we don't even know if it's handled on the client side. So for now, I removed `remoteValue` from `BookingTypeFilterUiState.kt`. de8fc3ee2a4466e3d7e349b6e13851653fc5b4a4

AdamGrzybkowski reviewed (commented):
I see that this PR closes the Linear issue, but to fully implement it, you need to implement the storage for the BookingType filter [here](https://github.com/woocommerce/woocommerce-android/blob/trunk/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/filter/data/BookingFilterRepository.kt). 

Are you planning on doing that separately?

irfano committed:
Use UiString for booking filter subtitles

Replaced the custom `BookingFilterListItemSubtitle` class with the generic `UiString` model to handle subtitle text in the bookings filter list.

irfano committed:
Remove unused remoteValue from BookingType

irfano committed:
Persist Booking Type filter in DataStore

This change introduces the functionality to save and retrieve the "Booking Type" filter to and from the DataStore.

irfano committed:
Exclude 'any' booking type from filter count

Do not increment the active filter count when the booking type filter is set to 'Any'.

AdamGrzybkowski reviewed (approved):
Looks good!
